tests/rest/admin/test_admin.py

```diff
@@ -685,3 +684,337 @@ def test_quarantine_all_media_by_user(self):
+        )
+
+        # Check that 3 rooms were returned
+        self.assertEqual(3, len(channel.json_body["rooms"]), msg=channel.json_body)
+
+        # Check their room_ids match
+        returned_room_ids = [room["room_id"] for room in channel.json_body["rooms"]]
+        common_room_ids = set(room_ids) & set(returned_room_ids)
+        self.assertEqual(
+            len(common_room_ids),
+            total_rooms,
+            msg="Different room_ids than expected returned. "
+            "Expected:\n%s\nReturned:\n%s" % (room_ids, returned_room_ids),
+        )
+
+        # Check that all fields are available
+        for r in channel.json_body["rooms"]:
+            self.assertIn("name", r)
+            self.assertIn("canonical_alias", r)
+            self.assertIn("joined_members", r)
+
+        # We shouldn't receive a next token here as there's no further rooms to show
+        self.assertTrue("next_token" not in channel.json_body)
+
+    def test_list_rooms_pagination(self):
+        """Test that we can get a full list of rooms through pagination"""
+        # Create 5 test rooms
+        total_rooms = 5
+        room_ids = []
+        for x in range(total_rooms):
+            room_id = self.helper.create_room_as(
+                self.admin_user, tok=self.admin_user_tok
+            )
+            room_ids.append(room_id)
+
+        returned_room_ids = []
+
+        # Request the list of rooms
+        start = 0
+        limit = 2
+
+        run_count = 0
+        should_repeat = True
+        while should_repeat:
+            run_count += 1
+
+            url = "/_synapse/admin/v1/rooms?from=%d&limit=%d" % (start, limit)
+            request, channel = self.make_request(
+                "GET", url.encode("ascii"), access_token=self.admin_user_tok,
+            )
+            self.render(request)
+            self.assertEqual(
+                200, int(channel.result["code"]), msg=channel.result["body"]
+            )
+
+            self.assertTrue("rooms" in channel.json_body)
+            for r in channel.json_body["rooms"]:
+                returned_room_ids.append(r["room_id"])
+
+            if "next_token" not in channel.json_body:
+                # We have reached the end of the list
+                should_repeat = False
+            else:
+                # Make another query with an updated start value
+                start = channel.json_body["next_token"]
+
+        # We should've queried the endpoint 3 times
+        self.assertEqual(
+            run_count,
+            3,
+            msg="Should've queried 3 times for 5 rooms with " "limit 2 per query",
+        )
+
+        # Check that we received all of the room ids
+        common_room_ids = set(room_ids) & set(returned_room_ids)
+        self.assertEqual(
+            len(common_room_ids),
+            total_rooms,
+            msg="Different room_ids than expected returned. "
+            "Expected:\n%s\nReturned:\n%s" % (room_ids, returned_room_ids),
```

Comment:
Again, probably want to pass in the actual collections rather than just checking lengths.